measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+            end = (
+                self.n_static_measurements
+                + (i - self.n_static_measurements + 1) * self.dynamic_Nt
+            )
+            # if no any time points from this DCM is selected, its installation cost should not be included
+            return m.if_install_dynamic[i] <= sum(
+                m.cov_y[j, j] for j in range(start, end)
+            )
+
+        # add constraints depending on if the FIM is defined as half triangle or all
+        if upper_diagonal_only:
+            m.total_fim_constraint = pyo.Constraint(m.dim_fim_half, rule=eval_fim)
         else:
-            # construct identity matrix
-            self.Sigma = np.identity(self.total_no_measure)
+            m.total_fim_constraint = pyo.Constraint(m.dim_fim, m.dim_fim, rule=eval_fim)
+            # only when the mixed-integer problem y are defined not only upper triangle
+            # this can help the performance of MIP problems so we keep it although not use it now
+            if mixed_integer:
+                m.sym = pyo.Constraint(m.n_responses, m.n_responses, rule=symmetry)
+
+        # if given fixed solution, no need to add the following constraints
+        if not fix:
+            # total dynamic timepoints number
+            m.total_number_dynamic_measurements = pyo.Var(
+                initialize=total_manual_num_init
+            )
+            # compute total dynamic timepoints number
+            m.manual = pyo.Constraint(rule=total_dynamic)
+            # total dynamic timepoints < total manual number
+            m.con_manual = pyo.Constraint(rule=total_dynamic_con)
+
+            ## integer cuts
+            # intiialize total number of measurements selected
+            m.total_number_measurements = pyo.Var(initialize=total_measure_initial)
+            # compute total number of measurements selected
+            m.integer_cut0 = pyo.Constraint(rule=integer_cut_0)
+            # let total number of measurements selected > 0, so we cut the all 0 solution
+            m.integer_cut0_in = pyo.Constraint(rule=integer_cut_0_ineq)
+
+            # relaxation constraints for y[a,b] = y[a]*y[b]
+            m.cov1 = pyo.Constraint(m.n_responses, m.n_responses, rule=y_covy1)
+            m.cov2 = pyo.Constraint(m.n_responses, m.n_responses, rule=y_covy2)
+            m.cov3 = pyo.Constraint(m.n_responses, m.n_responses, rule=y_covy3)
+
+            # dynamic-cost measurements installaction cost
+            def dynamic_install_init(m, j):
+                # if there is no installation cost
+                if dynamic_install_initial is None:
+                    return 0
+                # if there is installation cost
+                else:
+                    return dynamic_install_initial[j - self.n_static_measurements]
+
+            # we choose that if this is a mixed-integer problem, the dynamic installation flag is in {0,1}
+            if mixed_integer:
```

Comment:
Is this code reused in many places? If so, add it as a static method to the class. `var_domain = domain(mixed_integer)`. If you only use it in this function, you can declare it as a function inside this function.

Reply:
It's only used in this function; Defined it as a function